### PR TITLE
User Registration and Admin Updates

### PR DESCRIPTION
**Pull request by zkraime**
originally: #4718 and https://github.com/CenterForOpenScience/osf.io/pull/4796

### CODE REVIEW

**Comment by zkraime** [comment]
@sloria This should be good to go

File: admin/base/views.py [hunk 1]
```diff
@@ -5,14 +5,13 @@
...
 
-
 @login_required
 def home(request):
     context = {}
     return render(request, 'home.html', context)
 
 def password_reset_done(request, **kwargs):
-    messages.success(request, 'You have successfully reset your password and activated your admin account. Thank you')
+    messages.success(request, 'You have successfully reset your password and activated your admin account.')
```
Comment (Ghalko) [inline admin/base/views.py]: Change password is when a user is already logged in and wants to change their password. The hoops they should have to jump through should just be password length and repetition.

Reset password is when a user forgets their login information and they need a link or temporary password sent to their Email.

So if they are activating their account it shouldn't look as though they had to reset the password, and if they are resetting their password it shouldn't look like they are activating their account.

**Comment by Ghalko** [comment]
screen shot 2016-02-05 at 10 57 30 am

Occurs when clicking on the link after the registration email is sent and I am not logged in.

**Comment by zkraime** [comment]
@Ghalko Thanks for catching these, I'll address them right away

File: admin/templates/base.html [hunk 2]
```diff
@@ -82,6 +87,9 @@
             {% if 'prereg_group' in request.user.group_names %}
             <li><a href="{% url 'pre_reg:prereg' %}"><i class='fa fa-link'></i> <span>OSF Prereg</span></a></li>
             {% endif %}
+            {% if request.user.is_staff %}
+            <li><a href="{% url 'auth:register' %}"><span>User Registration</span></a></li>
```
Comment (Ghalko) [inline admin/templates/base.html]: I would suggest pre-pending `Admin` onto the `User Registration` in order to make it clear that it isn't the OSF. Also `fa-link`

Comment (zkraime) [inline admin/templates/base.html]: @Ghalko Roger that

**Comment by Ghalko** [comment]
I was able to change my password to a <5 character password.

**Comment by zkraime** [comment]
@Ghalko How did you do that?

**Comment by Ghalko** [comment]
1. Clicked on the change password link
2. typed in old password
3. typed in new 4 char password
4. retyped new 4 char password
4a. hit submit
5. logged out
6. logged back in using my new password

**Comment by zkraime** [comment]
@Ghalko I'm trying to reproduce using the change password link, but I'm getting a validation error for a 4 char password. 
What was the 4 char password?

**Comment by Ghalko** [comment]
A suggestion for the admin user registration page: redirect back to the same form page rather than taking them to their home page. It is a bit disconcerting, and if you have a second one or a 100th one to add (not that we will ever get that big, though maybe) then you don't have to click the menu link again.

**Comment by Ghalko** [comment]
@zkraime The password was `f1sh`

**Comment by zkraime** [comment]
@Ghalko I agree about redirecting back to the same, I had noticed that

**Comment by Ghalko** [comment]
At this moment the only thing I see that is really holding back this PR other than above issues is that you have a major problems with tabs and spaces. You should set your editor to use **4 spaces** in place of tabs.

**Comment by Ghalko** [comment]
@zkraime  Let me see if I need to restart my server or run migrate.

**Comment by Ghalko** [comment]
Have you done makemigrations for your changes? @zkraime

**Comment by zkraime** [comment]
@Ghalko I had before merging the other branch, but I needed to do it again; Ill push the migration files

**Comment by Ghalko** [comment]
@zkraime Still didn't fix the <5 char password. This time `cl4m`

**Comment by zkraime** [comment]
@Ghalko  I know, I can't reproduce. I just tried that and f1sh and I'm getting that it's too short. 
I'm not sure what the issue is and why it's not consistent. 
screen shot 2016-02-05 at 1 11 09 pm

**Comment by Ghalko** [comment]
@zkraime Unfortunately I am running from this same PR repo/branch. The commits I see here are the same ones I see on my environment. Is there something else that isn't getting committed, is setup dependent like local.py, or some other factor that would keep the change from taking effect on my end?

**Comment by zkraime** [comment]
defining AUTH_PASSWORD_VALIDATORS  in settings takes care of validating the password length so if you have that, I don't see the cause of the problem

**Comment by Ghalko** [comment]
I am pinging Steve with this PR.

**Comment by Ghalko** [comment]
@zkraime I can update this if it is out of your scope for now.

**Comment by zkraime** [comment]
@Ghalko Sorry for the delay; I'm coming in Monday and Tuesday and will finish updates to this and then finish up the Conferences PR. Thanks

File: admin/base/urls.py [hunk 3]
```diff
@@ -20,8 +21,17 @@
                          url(r'^prereg/', include('admin.pre_reg.urls', namespace='pre_reg')),
                          url(r'^reset/(?P<uidb64>[0-9A-Za-z_\-]+)/(?P<token>[0-9A-Za-z]{1,13}-[0-9A-Za-z]{1,20})/$',
                             views.password_reset_confirm_custom, name='password_reset_confirm'),
-                         url(r'^reset/done/$', views.password_reset_done,
+                         url(r'^resetpassword/$', 'django.contrib.auth.views.password_reset', name='reset_password'),
+                         url(r'^password_reset_done/$', 'django.contrib.auth.views.password_reset_done', name='password_reset_done'),
+                         url(r'^password/reset/complete/$',
```
Comment (sloria) [inline admin/base/urls.py]: Try to keep your URLs consistent. Something along the lines of:

```
/accounts/password_reset/ -> password_reset
/accounts/password_reset/done/ -> passwor_reset_done
/accounts/password_reset/complete/ -> password_reset_completed
#....
```

**Comment by sloria** [comment]
Are there any tests for these changes? @Ghalko

**Comment by sloria** [comment]
Just a few drive-by comments. For now, see if you can clean up the URLs and maybe add some tests. Would like to merge this one soon, since this is a dependency of the Spam PRs.

**Comment by whit537** [comment]
@sloria Is this a good one for me to look at this week? The [review queue on Jira](https://openscience.atlassian.net/issues/?filter=11402) is empty at the moment.

**Comment by sloria** [comment]
@whit537 Yes, go ahead with this one. After this one, there's this roadmap feature: https://github.com/CenterForOpenScience/osf.io/pull/5183 (currently not ready for review, but might be ready by the time you've finished with this PR).

**Comment by whit537** [comment]
@sloria Cool, will do.

**Comment by whit537** [comment]
What's the Jira ticket for this?

**Comment by whit537** [comment]
Here it is: https://openscience.atlassian.net/browse/ADM-47.

**Comment by whit537** [comment]
Lots of different things going on on this PR ...

**Comment by whit537** [comment]
Looks like this already got some review in #4796 and #4718.

**Comment by whit537** [comment]
##### Punchlist—for @zkraime 

 - [ ] make Travis happy

##### Investigations—for @whit537

 - [ ] understand what this PR is trying to do

File: admin/base/settings/defaults.py [hunk 4]
```diff
@@ -23,9 +25,30 @@
...
+            'min_length': 5,
+        }
+    },
+]
+
 # Email settings. Account created for testing. Password shouldn't be hardcoded
+# [DEVOPS] this should be set to 'django.core.mail.backends.smtp.EmailBackend' in the > dev local.py.
 EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'
 
+# Sendgrip Email Settings - Using OSF credentials
```
Comment (whit537) [inline admin/base/settings/defaults.py]: grip &rarr; grid

**Comment by whit537** [comment]
P.S. @sloria This seems like a great example of where smaller batch size would help keep work flowing more smoothly.

**Comment by whit537** [comment]
The purpose stated at https://github.com/CenterForOpenScience/osf.io/pull/4718#issue-121661618 matches the list at [ADM-47](https://openscience.atlassian.net/browse/ADM-47), so that seems like a likely candidate for describing what this PR is about.

**Comment by whit537** [comment]
But there's a separate purpose at https://github.com/CenterForOpenScience/osf.io/pull/4796#issue-123025759. How does #4796 relate to this PR?

**Comment by whit537** [comment]
They both share a6d27826a247615b5de4838d76bf27b714aa144f, but their histories seem otherwise unrelated, for some reason (and why does GitHub report the commit comments only on the one PR and not the other?).

screen shot 2016-03-14 at 3 20 32 pm
screen shot 2016-03-14 at 3 20 48 pm

**Comment by whit537** [comment]
Looks like they sync up by the end, though:

screen shot 2016-03-14 at 3 29 34 pm

--------------------

screen shot 2016-03-14 at 3 29 44 pm

**Comment by whit537** [comment]
Well, I guess we should bring in the purpose from #4796 as well, then.

**Comment by whit537** [comment]
I'm going back through my notes at https://github.com/CenterForOpenScience/osf.io/pull/5040#issuecomment-186386144 to stand up an admin instance again.

**Comment by whit537** [comment]
/me discovers an admin [README](https://github.com/CenterForOpenScience/osf.io/blob/286688ff9e98455787425de4877880578bf3d1f7/admin/README.md) ...

**Comment by whit537** [comment]
README added recently, in #5129.

**Comment by whit537** [comment]
Alright, I'm in again. Now ... what am I looking for?

**Comment by whit537** [comment]
[These](https://github.com/CenterForOpenScience/osf.io/pull/4992#issuecomment-196342715) UI errors look pretty straightforward, let's start with those ...

**Comment by whit537** [comment]
> Restrict home page by authenticated (fix "Logged in as Sign Out" bug)

Figured I'd start with this because I noticed it on `develop` as I was getting set up. I still see it on this PR, tho. :-/

**Comment by whit537** [comment]
Here's the list of changed files. Which files relate to which of the 11 or 12 purposes on this PR?

```
admin/base/settings/defaults.py
admin/base/urls.py
admin/base/views.py
admin/common_auth/admin.py
admin/common_auth/forms.py
admin/common_auth/migrations/0002_auto_20160205_1157.py
admin/common_auth/models.py
admin/common_auth/urls.py
admin/common_auth/views.py
admin/pre_reg/utils.py
admin/templates/base.html
admin/templates/common_auth/emails/account_creation_subject.txt
admin/templates/common_auth/emails/invitation_email.html
admin/templates/emails/account_creation_subject.txt
admin/templates/home.html
admin/templates/login.html
admin/templates/register.html
admin/templates/registration/password_change_done.html
admin/templates/registration/password_reset_email.html
admin_tests/common_auth/__init__.py
admin_tests/common_auth/test_views.py
celerybeat-schedule
tasks/__init__.py
```

**Comment by whit537** [comment]
| Purpose | Files |
|:--------------|:-------------|
| Custom email templates |  |
| Password length restriction |  |
| UI Errors - Login view cutoff |  |
| UI Errors - Restrict home page by authenticated (fix "Logged in as Sign Out" bug) |  |
| UI Errors - Add password change page for users |  |
| UI Errors - Add forgot password link |  |
| UI Errors - Display proper Django message class type (current 'patch' handles only success and error) |  |
| UI Errors - Handle warning/failure cases for send_email_invitation in Django-admin |  |
| Set up OSF email server for admin app |  |
| Needs to be tested on staging |  |
| Add basic registration fields (email, first/last name, password) |  |
| Add FilterSelectMultiple widget for displaying/selecting/assigning Permissions (and Groups?) |  |
| Add email invitation upon new registration on custom page |  |

```
admin/base/settings/defaults.py
admin/base/urls.py
admin/base/views.py
admin/common_auth/admin.py
admin/common_auth/forms.py
admin/common_auth/migrations/0002_auto_20160205_1157.py
admin/common_auth/models.py
admin/common_auth/urls.py
admin/common_auth/views.py
admin/pre_reg/utils.py
admin/templates/base.html
admin/templates/common_auth/emails/account_creation_subject.txt
admin/templates/common_auth/emails/invitation_email.html
admin/templates/emails/account_creation_subject.txt
admin/templates/home.html
admin/templates/login.html
admin/templates/register.html
admin/templates/registration/password_change_done.html
admin/templates/registration/password_reset_email.html
admin_tests/common_auth/__init__.py
admin_tests/common_auth/test_views.py
celerybeat-schedule
tasks/__init__.py
```

**Comment by whit537** [comment]
What does "Custom email templates" mean?

**Comment by whit537** [comment]
Is that, like, "*start* using custom email templates"? What were we using before? What could *non*-custom email templates mean in our context here?

File: admin/templates/common_auth/emails/invitation_email.html [hunk 5]
```diff
@@ -0,0 +1 @@
+{% extends 'registration/password_reset_email.html' %}
```
Comment (whit537) [inline admin/templates/common_auth/emails/invitation_email.html]: What's the purpose of using an `extends` here? Why not have the template code directly in this file? I don't find `password_reset_email` used anywhere else in the codebase, and I don't see any other files in `admin/templates/common_auth/emails/` that would establish a precedent.

**Comment by whit537** [comment]
I went through the password reset flow locally and it works afaict.

File: admin/base/settings/defaults.py [hunk 6]
```diff
@@ -23,9 +25,30 @@
...
 ]
 
+AUTH_PASSWORD_VALIDATORS = [
+    {
+        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
+    },
+    {
+        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
+        'OPTIONS': {
+            'min_length': 5,
```
Comment (whit537) [inline admin/base/settings/defaults.py]: This is not having the desired effect that I can see. I am able to reproduce @Ghalko's result at https://github.com/CenterForOpenScience/osf.io/pull/4992#issuecomment-180451300.

Comment (zkraime) [inline admin/base/settings/defaults.py]: @whit537 I should have added a comment.
This only works in Django 1.9, develop currently has 1.8

**Comment by zkraime** [comment]
@whit537 Thanks for the review, I will address the rest next time I'm in the office which will be tomorrow

**Comment by whit537** [comment]
@zkraime I'm just getting started. :-) I'll update the [punchlist](https://github.com/CenterForOpenScience/osf.io/pull/4992#issuecomment-196342715) as I work through items under "Investigations."

File: admin/base/settings/defaults.py (hunk 6 above)
Comment (whit537) [inline admin/base/settings/defaults.py]: @zkraime Okay, so then we should either:

 1. drop the "Password length restriction" requirement from the [purpose](https://github.com/CenterForOpenScience/osf.io/pull/4992#issuecomment-196511342) of this PR
 1. meet the "Password length restriction" requirement another way
 1. upgrade to Django 1.9

Since this PR is so mondo already, I recommend (1). :-)

**Comment by whit537** [comment]
I've got my environment in an inconsistent state. I'm trying to run locally and getting an `AttributeError` when signing in. Also spent some time learning about `reset` and writing a `kill-osf.sh` script to work around kludges with this `Procfile` setup I'm building up.

screen shot 2016-03-18 at 9 51 31 am

**Comment by whit537** [comment]
Okay, figured it out (configuration mismatch between web app and fakecas). Back on track and proceeding with review ...

**Comment by whit537** [comment]
Of course, this PR is for `admin` and not `website` anyway ... :mouse:

File: admin/pre_reg/utils.py [hunk 7]
```diff
@@ -5,10 +5,14 @@
...
 
 def get_prereg_reviewers():
-    return MyUser.objects.filter(
-        groups__name='prereg_group'
-    ).annotate(
-        fuller_name=Concat('first_name', Value(' '), 'last_name')
-    ).values_list(
-        'email', 'fuller_name'
-    )
+    # Note - fixes django.db.utils.OperationalError: no such table' error if ever one erases a db table and tries to remigrate
```
Comment (whit537) [inline admin/pre_reg/utils.py]: Gosh, why would we need to plan for that? Seems like if you're dropping tables then it's up to you to deal with it on your own. Is there a use case I'm missing here?

File: admin/pre_reg/utils.py [hunk 8]
```diff
@@ -5,10 +5,14 @@
...
+    try:
+        return MyUser.objects.filter(
+            groups__name='prereg_group'
+        ).annotate(
+            fuller_name=Concat('first_name', Value(' '), 'last_name')
+        ).values_list(
+            'email', 'fuller_name'
+        )
+    except Exception:
```
Comment (whit537) [inline admin/pre_reg/utils.py]: Even if this error handling is legitimate, bare `except`s are bad because they hide legitimate errors.

File: tasks/__init__.py [hunk 9]
```diff
@@ -85,6 +85,11 @@ def adminserver(port=8001):
     cmd = '{} python manage.py runserver {} --nothreading'.format(env, port)
     run(cmd, echo=True, pty=True)
 
+@task
+def admin_task(args):
```
Comment (whit537) [inline tasks/__init__.py]: This is redundant with `inv admin.manage`, a task the existence of which is [hinted](https://github.com/CenterForOpenScience/osf.io/tree/efb51ffac31fe79a0ad508dded1764e3100bf8fd/admin#how-to-run-the-admin-interface) at in the README.

File: admin/base/settings/defaults.py [hunk 10]
```diff
@@ -23,9 +25,31 @@
...
+
 # Email settings. Account created for testing. Password shouldn't be hardcoded
+# [DEVOPS] this should be set to 'django.core.mail.backends.smtp.EmailBackend' in the > dev local.py.
 EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'
 
+# Sendgrid Email Settings - Using OSF credentials
+
+EMAIL_HOST = 'smtp.sendgrid.net'
+EMAIL_HOST_USER = 'osf-smtp'
+EMAIL_HOST_PASSWORD = osf_settings.MAIL_PASSWORD  # local.py
```
Comment (whit537) [inline admin/base/settings/defaults.py]: If we're going to delegate the password to `website.settings`, let's do the same with host and username.

**Comment by whit537** [comment]
The new admin user registration page seems to work. It's a little weird that we send a password reset instead of a "Welcome!" email, but I guess that's an MVP and this is just for internal users anyway.

**Comment by whit537** [comment]
>   - [ ] Handle warning/failure cases for send_email_invitation in Django-admin

This [appears](https://github.com/CenterForOpenScience/osf.io/blob/efb51ffac31fe79a0ad508dded1764e3100bf8fd/admin/common_auth/admin.py#L29) not to have been done—which is fine! Another way to shed weight from this PR. :-)

**Comment by whit537** [comment]
>> Restrict home page by authenticated (fix "Logged in as Sign Out" bug)
>
Figured I'd start with this because I noticed it on `develop` as I was getting set up. I still see it on this PR, tho. :-/

Let's drop this too, I guess.

File: admin/base/urls.py [hunk 11]
```diff
@@ -22,10 +23,19 @@
...
                              include('admin.users.urls', namespace='users')),
                          url(r'^prereg/', include('admin.pre_reg.urls', namespace='pre_reg')),
-                         url(r'^reset/(?P<uidb64>[0-9A-Za-z_\-]+)/(?P<token>[0-9A-Za-z]{1,13}-[0-9A-Za-z]{1,20})/$',
+                         url(r'^accounts/password_reset/$', 'django.contrib.auth.views.password_reset', name='reset_password'),
+                         url(r'^accounts/password_reset/confirm/(?P<uidb64>[0-9A-Za-z_\-]+)/(?P<token>[0-9A-Za-z]{1,13}-[0-9A-Za-z]{1,20})/$',
                             views.password_reset_confirm_custom, name='password_reset_confirm'),
-                         url(r'^reset/done/$', views.password_reset_done,
+                         url(r'^accounts/password_reset/done/$', 'django.contrib.auth.views.password_reset_done', name='password_reset_done'),
+                         url(r'^accounts/password_reset/complete/$',
+                            'django.contrib.auth.views.password_reset_complete',
```
Comment (whit537) [inline admin/base/urls.py]: Golly. What might be the difference between `password_reset_done` and `password_reset_complete`?

**Comment by whit537** [comment]
Alright, so I'm starting to wrap my head around this PR. I'm seeing three basic purposes here:


| Basic Purpose | Purposes on https://github.com/CenterForOpenScience/osf.io/pull/4992#issuecomment-196511342 |
|:----------|:----------------|
| implement password changing | - Add password change page for users |
| implement password resetting | - Add forgot password link |
| implement user registration | - Custom email templates<br>- Display proper Django message class type (current 'patch' handles only success and error)<br> - Handle warning/failure cases for send_email_invitation in Django-admin<br>- Add basic registration fields (email, first/last name, password)<br>- Add FilterSelectMultiple widget for displaying/selecting/assigning Permissions (and Groups?)<br>- Add email invitation upon new registration on custom page |
| extras | - Password length restriction<br>- Login view cutoff<br>- Restrict home page by authenticated (fix "Logged in as Sign Out" bug) |
| null | - Set up OSF email server for admin app<br>- Needs to be tested on staging |

**Comment by whit537** [comment]
@zkraime In the future, please scope your PRs more tightly. This PR probably should've been four or more separate PRs (per https://github.com/CenterForOpenScience/osf.io/pull/4992#issuecomment-199334873). Smaller, tightly focused PRs are easier to move through the pipeline. Thanks. :-)

File: admin/base/urls.py [hunk 12]
```diff
@@ -22,10 +23,19 @@
                          url(r'^users/',
                              include('admin.users.urls', namespace='users')),
                          url(r'^prereg/', include('admin.pre_reg.urls', namespace='pre_reg')),
-                         url(r'^reset/(?P<uidb64>[0-9A-Za-z_\-]+)/(?P<token>[0-9A-Za-z]{1,13}-[0-9A-Za-z]{1,20})/$',
+                         url(r'^accounts/password_reset/$', 'django.contrib.auth.views.password_reset', name='reset_password'),
+                         url(r'^accounts/password_reset/confirm/(?P<uidb64>[0-9A-Za-z_\-]+)/(?P<token>[0-9A-Za-z]{1,13}-[0-9A-Za-z]{1,20})/$',
```
Comment (whit537) [inline admin/base/urls.py]: What is this magic regex? Let's move it out into a variable with a comment.

File: admin/base/urls.py (hunk 11 above)
Comment (whit537) [inline admin/base/urls.py]: > The page shown after a user has been emailed a link to reset their password.

https://docs.djangoproject.com/en/1.8/topics/auth/default/#django.contrib.auth.views.password_reset_done

> Presents a view which informs the user that the password has been successfully changed.

https://docs.djangoproject.com/en/1.8/topics/auth/default/#django.contrib.auth.views.password_reset_complete

File: admin/base/views.py [hunk 13]
```diff
@@ -17,7 +17,7 @@ def home(request):
 
 
 def password_reset_done(request, **kwargs):
```
Comment (whit537) [inline admin/base/views.py]: I believe this view function can be removed entirely, because we're no longer using it in `admin.base.urls`.

File: admin/common_auth/views.py [hunk 14]
```diff
@@ -21,6 +24,47 @@ def login(request):
...
+            else:
+                print(form.errors)
+                context = {'form': form}
+                return render(request, 'register.html', context, status=400)
+        else:
+            reg_form = CustomUserRegistrationForm()
+            context = {'form': reg_form}
+            return render(request, 'register.html', context)
+    else:
+        # Best way to inform user: pop-up alert or display message after redirect?
```
Comment (whit537) [inline admin/common_auth/views.py]: Looks like we're implementing the second option, which is fine. We can remove the comment.

File: admin/common_auth/views.py [hunk 15]
```diff
@@ -21,6 +24,47 @@ def login(request):
...
     return render(request, 'login.html', context)
 
+
 def logout(request):
     logout_user(request)
     return redirect('auth:login')
+
+
+# Permission restriction based on staff (only staff can register new users). Use is_staff boolean?
+def register(request):
```
Comment (whit537) [inline admin/common_auth/views.py]: Since we're already returning from multiple places in this function, let's restructure the function to avoid such deep nesting, using a "gating" pattern:

```python
def register(request):
    if not request.user.is_staff:
        return redirect('auth:login')

    if not request.method != 'POST':
        return render(request, 'register.html', context)

    form = CustomUserRegistrationForm(request.POST)
    if not form.is_valid():
        return render(request, 'register.html', context, status=400)

    pass # process form
    return redirect('auth:register')
```

File: admin/common_auth/views.py [hunk 16]
```diff
@@ -21,6 +24,47 @@ def login(request):
...
+
+                new_user.save()
+                reset_form = PasswordResetForm({'email': new_user.email}, request.POST)
+                assert reset_form.is_valid()
+                reset_form.save(
+                    subject_template_name='common_auth/emails/account_creation_subject.txt',
+                    email_template_name='common_auth/emails/invitation_email.html',
+                    request=request
+                )
+                messages.success(request, 'Registration successful')  # add email reference here
```
Comment (whit537) [inline admin/common_auth/views.py]: > add email reference here

This is a TODO item, yes?

File: admin/common_auth/views.py [hunk 17]
```diff
@@ -21,6 +24,47 @@ def login(request):
     context = {'form': form}
     return render(request, 'login.html', context)
 
+
 def logout(request):
     logout_user(request)
     return redirect('auth:login')
+
+
+# Permission restriction based on staff (only staff can register new users). Use is_staff boolean?
```
Comment (whit537) [inline admin/common_auth/views.py]: We are using the is_staff boolean. We can drop the question from this comment.

File: admin/common_auth/views.py [hunk 18]
```diff
@@ -21,6 +24,47 @@ def login(request):
...
+def register(request):
+    if request.user.is_staff:
+        if request.method == 'POST':
+            form = CustomUserRegistrationForm(request.POST)
+            if form.is_valid():
+                email = form.cleaned_data['email']
+                password = form.cleaned_data['password1']
```
Comment (whit537) [inline admin/common_auth/views.py]: I checked to make sure that password confirmation is at least being performed on the client side.

screen shot 2016-03-21 at 1 45 02 pm

File: admin/templates/common_auth/emails/invitation_email.html (hunk 5 above)
Comment (whit537) [inline admin/templates/common_auth/emails/invitation_email.html]: Perhaps the intention is to customize this form in future PRs (cf. https://github.com/CenterForOpenScience/osf.io/pull/4992#issuecomment-199289211)?

File: admin/common_auth/views.py [hunk 19]
```diff
@@ -21,6 +24,47 @@ def login(request):
...
+                new_user = MyUser.objects.create_user(email=email, password=password)
+                new_user.first_name = form.cleaned_data['first_name']
+                new_user.last_name = form.cleaned_data['last_name']
+                group_perms = form.cleaned_data['group_perms']
+
+                for group in group_perms:
+                    new_user.groups.add(group)
```
Comment (whit537) [inline admin/common_auth/views.py]: How do I test this functionality? I see an empty select for "Select account Permissions," which is what I would expect to influence this.

screen shot 2016-03-21 at 1 58 44 pm

File: admin_tests/common_auth/test_views.py [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+        # Every field is required, password length should be >= 5 (need Django 1.9 to use AUTH_PASSWORD_VALIDATORS)
+        form_data = {
+            'email': '',
+            'first_name': '',
+            'last_name': '',
+            'password1': 'pass',
+            'password2': 'pass',
+        }
+        form = CustomUserRegistrationForm(form_data)
+        self.assertFalse(form.is_valid())
```
Comment (whit537) [inline admin_tests/common_auth/test_views.py]: How about at least checking for the number of errors on the form, since otherwise it's not clear from this test which fields are failing since `is_valid` is so coarse.

File: admin/common_auth/admin.py [hunk 21]
```diff
@@ -11,7 +11,7 @@ class PermissionAdmin(admin.ModelAdmin):
 
 
 class CustomUserAdmin(UserAdmin):
-    add_form = CustomUserRegistrationForm
+    add_form = UserCreationForm
```
Comment (whit537) [inline admin/common_auth/admin.py]: Why are we no longer using `CustomUserRegistrationForm` here? I'm surprised that it still exists even though we're removing it here.

**Comment by whit537** [comment]
Okay! First pass completed, @zkraime, and [punchlist](https://github.com/CenterForOpenScience/osf.io/pull/4992#issuecomment-196342715) updated! :basketball: is in your court! :-)

File: admin/common_auth/views.py (hunk 19 above)
Comment (Ghalko) [inline admin/common_auth/views.py]: Add a group.

Comment (whit537) [inline admin/common_auth/views.py]: That happens in the "real" Django Admin, yes? How do I access that, I wonder ... ?

Comment (whit537) [inline admin/common_auth/views.py]: Found it ...

Comment (whit537) [inline admin/common_auth/views.py]: I added a group, and I still get an empty permissions list on the user registration page.

screen shot 2016-03-21 at 4 18 51 pm

screen shot 2016-03-21 at 4 19 33 pm

Comment (zkraime) [inline admin/common_auth/views.py]: For now the form is hardcoded to use the prereg_group, 
but the idea is to have it generate dynamically (for now we could display all() groups or determine which ones some other way)

File: admin_tests/common_auth/test_views.py (hunk 20 above)
Comment (zkraime) [inline admin_tests/common_auth/test_views.py]: Good point

File: admin/templates/common_auth/emails/invitation_email.html (hunk 5 above)
Comment (zkraime) [inline admin/templates/common_auth/emails/invitation_email.html]: Sorry for the confusion - I believe I was going to extend the template further but the base customized one is now under admin/templates/emails/password_reset_email.html

**Comment by zkraime** [comment]
Thanks @whit537, I will notify you when I'm finished responding to the checklist items

**Comment by whit537** [comment]
Sounds good, @zkraime. :)

**Comment by brianjgeiger** [comment]
@Ghalko Could you look into the failing test? I think it's in the post-merge-with-develop branch of the testing, but I'm not positive. Might just need to merge develop into this branch and verify that it still passes.

File: admin/base/settings/defaults.py (hunk 6 above)
Comment (zkraime) [inline admin/base/settings/defaults.py]: Ok I think we will drop it from this PR and add it in another

File: admin/pre_reg/utils.py (hunk 8 above)
Comment (zkraime) [inline admin/pre_reg/utils.py]: @cwisecarver What do you suggest?

File: admin/base/urls.py (hunk 12 above)
Comment (zkraime) [inline admin/base/urls.py]: These regexes (well, the entire url) come directly from django, they're used for the password reset flow

File: admin/common_auth/views.py (hunk 16 above)
Comment (zkraime) [inline admin/common_auth/views.py]: It was never an officially documented item, but I figured it would be good to add to the user feedback message

File: admin/pre_reg/utils.py (hunk 7 above)
Comment (cwisecarver) [inline admin/pre_reg/utils.py]: :+1:

File: admin/pre_reg/utils.py (hunk 8 above)
Comment (cwisecarver) [inline admin/pre_reg/utils.py]: @zkraime It depends on what exceptions you're trying to account for. django ORM's filter is pretty forgiving. If your query doesn't return any results you should just get an empty queryset with no exception. If your `Group` model didn't have a `name` you would get a [FieldError](https://docs.djangoproject.com/en/1.9/ref/exceptions/#fielderror). Same thing if `email` or `fuller_name` weren't fields. 

Generally, when I'm putting in a try except I do it because I got an exception while functional testing and it was an exception I knew *could* happen to a user but I wanted to handle that in some way in order to explain it to them and not just give them an error without recourse. Chances are if you're querying on fields that don't exist, errything broke and messaging to the user isn't going to fix it. If you're trying to account for invalid user input, that should probably be done at the form.

Comment (zkraime) [inline admin/pre_reg/utils.py]: Let's move this to another PR?

File: admin/base/settings/defaults.py (hunk 6 above)
Comment (Ghalko) [inline admin/base/settings/defaults.py]: 1.9 is now standard on develop.

**Comment by sloria** [comment]
@Ghalko Is this PR still needed, or should it be closed?
